pkg/status/combinedstatus-resolution.go

```diff
@@ -0,0 +1,308 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package status
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog/v2"
+
+	"github.com/kubestellar/kubestellar/api/control/v1alpha1"
+	"github.com/kubestellar/kubestellar/pkg/util"
+)
+
+// combinedStatusResolution is a struct that represents the resolution of a
+// combinedstatus. A combinedstatus resolution maps to (binding, object).
+// A resolution's logic is as follows:
+// - Composition:
+//   - A combinedstatus resolution is composed of multiple statuscollector
+//     data.
+//   - statuscollector data is composed of collector-clauses. Each
+//     collector-clause has an expression that is evaluated against every
+//     workstatus that is associated with the statuscollector, and the result
+//     is cached.
+//
+// - Flow:
+//   - When a statuscollector is added to the combinedstatus resolution,
+//     every clause in the statuscollector is evaluated against every workstatus
+//     that is associated with the statuscollector.
+//   - When a workstatus is added to / modified in the combinedstatus
+//     resolution, every clause in every statuscollector is evaluated against
+//     the workstatus. The caching of collector-clause -> evaluation makes
+//     calculating combinedFields efficient if present.
+//
+// - Ownership:
+//   - A combinedstatus resolution is owned by the binding that it is
+//     associated with.
+type combinedStatusResolution struct {
+	sync.RWMutex
+
+	// statusCollectorNameToData is a map of status collector names to
+	// their corresponding data.
+	// If a status collector name is mapped to nil, it means that the status
+	// collector has not been processed.
+	statusCollectorNameToData map[string]*statusCollectorData
+	// workStatusToStatusCollectorName is a map of workstatus references to
+	// a set of statuscollector names that they contribute to.
+	workStatusToStatusCollectorName map[util.ObjectIdentifier]sets.Set[string]
+
+	// ownerReference identifies the binding that this resolution is
+	// associated with as an owning object.
+	ownerReference *metav1.OwnerReference
+}
+
+// statusCollectorData is a struct that represents the data of a status
+// collector in a combinedstatus resolution.
+type statusCollectorData struct {
+	*statusCollectorClauses
+
+	// workStatusRefToData is a map of workstatus references to their
+	// corresponding work status data.
+	// A workstatus reference is identical to its source object's identifier,
+	// apart from the namespace, which is set to the associated clusterName.
+	// If a workstatus reference is mapped to nil, it means that the workstatus
+	// has not been evaluated against the statuscollector's clauses.
+	workStatusRefToData map[util.ObjectIdentifier]workStatusData
```

Comment:
Each statusCollectorData is specific to one workload object. All that is needed in the domain of this map is the WEC name (equivalently, mailbox namespace name).